spec.bs

```diff
@@ -289,46 +318,180 @@ To <dfn for="PrivateAggregation">obtain an aggregatable report</dfn> given an [=
 
-Issue: Don't we need to take into account the time "now" as well?
+            Note: On startup, it is possible the user agent will need to send many reports whose report times passed while the browser was closed. Adding random delay prevents temporal joining of reports.
+        1. Wait until the [=current wall time=] is equal to or after |report|'s [=aggregatable report/report time=].
+        1. Optionally, wait a further [=implementation-defined=] non-negative [=duration=].
 
-Sending reports {#sending-reports}
-----------------------------------
+            Note: This is intended to allow user agents to optimize device resource usage.
+        1. Run [=attempt to deliver a report=] with |report|.
+    1. [=If aborted=], set |report|'s [=aggregatable report/queued=] value to false.
+
+        Note: It might be more practical to perform these steps when the [=user agent=] next starts up.
+
+To <dfn>attempt to deliver a report</dfn> given an [=aggregatable report=] |report|:
+1. Let |url| be the result of [=obtaining a reporting endpoint=] given |report|'s [=aggregatable report/reporting origin=] and |report|'s [=aggregatable report/api=].
+1. Let |data| be the result of [=serializing an aggregatable report=] given |report|.
+1. If |data| is an error, [=list/remove=] |report| from the [=aggregatable report cache=]. (TODO: do we need to queue this task?)
+1. Let |request| be the result of [=creating a report request=] given |url| and |data|.
+1. [=Queue a task=] to [=fetch=] |request| with [=fetch/processResponse=] being the following steps:
+    1. Let |shouldRetry| be an [=implementation-defined=] [=boolean=]. The value should be false if no error occurred.
+    1. If |shouldRetry| is true, set |report|'s [=aggregatable report/queued=] value to false.
+    1. Otherwise, [=list/remove=] |report| from the [=aggregatable report cache=].
+
+To <dfn>obtain a reporting endpoint</dfn> given an [=origin=] |reportingOrigin| and [=context type=] |api|, perform the following steps. They return a [=URL=].
+1. [=Assert=]: |reportingOrigin| is a [=potentially trustworthy origin=].
+1. Let |path| be the [=string/concatenation=] of «"<code>.well-known/private-aggregation/report-</code>", |api|».
+1. Let |base| be the result on running the [=URL parser=] on the [=serialization of an origin|serialization=] of |reportingOrigin|.
+1. [=Assert=]: |base| is not failure.
+1. Let |result| be the result of running the [=URL parser=] on |path| with |base|.
+1. [=Assert=]: |result| is not failure.
+1. Return |result|.
+
+To <dfn>create a report request</dfn> given a [=URL=] |url| and a [=byte sequence=] |body|:
+1. Let |request| be a new [=request=] with the following properties:
+    :   [=request/method=]
+    ::  "`POST`"
+    :   [=request/URL=]
+    ::  |url|
+    :   [=request/header list=]
+    ::  «("`Content-Type`", "`application/json`")»
+    :   [=request/unsafe-request flag=]
+    ::  set
+    :   [=request/body=]
+    ::  |body|
```

Comment:
ARA uses "A [body](https://fetch.spec.whatwg.org/#concept-body) whose [source](https://fetch.spec.whatwg.org/#concept-body-source) is body.", shall we aligh?

Reply:
The fetch spec defines the body as null, a byte sequence or a body, so I'd lean slightly towards just passing the byte sequence in and letting the fetch spec do the rest. Wdyt?